pintracker/stateless/stateless.go

```diff
@@ -0,0 +1,490 @@
+	// get ipfs status for all
+	localpis, err := spt.localStatus(false)
+	if err != nil {
+		logger.Error(err)
+		return nil, err
+	}
+
+	for _, p := range spt.optracker.FilterOps(optracker.OperationPin, optracker.PhaseError) {
+		if _, ok := localpis[p.Cid().String()]; ok {
+			spt.optracker.Clean(p)
+		}
+	}
+
+	for _, p := range spt.optracker.FilterOps(optracker.OperationUnpin, optracker.PhaseError) {
+		if _, ok := localpis[p.Cid().String()]; ok {
+			spt.optracker.Clean(p)
+		}
+	}
+
+	return spt.getErrorsAll(), nil
+}
+
+// Sync returns the updated local status for the given Cid.
+func (spt *Tracker) Sync(c *cid.Cid) (api.PinInfo, error) {
+	return spt.Status(c), nil
+}
+
+// RecoverAll attempts to recover all items tracked by this peer.
+func (spt *Tracker) RecoverAll() ([]api.PinInfo, error) {
+	statuses := spt.StatusAll()
+	resp := make([]api.PinInfo, 0)
+	for _, st := range statuses {
+		r, err := spt.Recover(st.Cid)
+		if err != nil {
+			return resp, err
+		}
+		resp = append(resp, r)
+	}
+	return resp, nil
+}
+
+// Recover will re-track or re-untrack a Cid in error state,
+// possibly retriggering an IPFS pinning operation and returning
+// only when it is done.
+func (spt *Tracker) Recover(c *cid.Cid) (api.PinInfo, error) {
+	logger.Infof("Attempting to recover %s", c)
+	pInfo, ok := spt.getError(c)
+	if !ok {
+		return spt.Status(c), nil
+	}
+
+	var err error
+	switch pInfo.Status {
+	case api.TrackerStatusPinError:
+		err = spt.enqueue(api.PinCid(c), optracker.OperationPin)
+	case api.TrackerStatusUnpinError:
+		err = spt.enqueue(api.PinCid(c), optracker.OperationUnpin)
+	}
+	if err != nil {
+		return spt.Status(c), err
+	}
+
+	pi, ok := spt.optracker.GetExists(c)
+	if !ok {
+		return spt.Status(c), nil
+	}
+
+	return pi, nil
+}
+
+func (spt *Tracker) ipfsStatusAll() (map[string]api.PinInfo, error) {
+	var ipsMap map[string]api.IPFSPinStatus
+	err := spt.rpcClient.Call(
+		"",
+		"Cluster",
+		"IPFSPinLs",
+		"recursive",
+		&ipsMap,
+	)
+	if err != nil {
+		logger.Error(err)
+		return nil, err
+	}
+	pins := make(map[string]api.PinInfo, 0)
+	for cidstr, ips := range ipsMap {
+		c, err := cid.Decode(cidstr)
+		if err != nil {
+			logger.Error(err)
+			continue
+		}
+		p := api.PinInfo{
+			Cid:    c,
+			Peer:   spt.peerID,
+			Status: ips.ToTrackerStatus(),
+		}
+		pins[cidstr] = p
+	}
+	return pins, nil
+}
+
+// localStatus returns a joint set of consensusState and ipfsStatus
+// marking pins which should be remote and leaving any ipfs pins that
+// aren't in the consensusState out.
+func (spt *Tracker) localStatus(incRemote bool) (map[string]api.PinInfo, error) {
+	pininfos := make(map[string]api.PinInfo)
+
+	// get shared state
+	var csps []api.PinSerial
+	err := spt.rpcClient.Call(
+		"",
+		"Cluster",
+		"Pins",
+		struct{}{},
+		&csps,
+	)
+	if err != nil {
+		logger.Error(err)
+		return nil, err
+	}
+	var cspis []api.Pin
+	for _, p := range csps {
+		cspis = append(cspis, p.ToPin())
```

Comment:
hmm, the fact that the whole state needs to be serialized, only to re-convert it to the original form supports my idea that we could pass the state component directly to this pintracker during initialization and work directly with it.

Reply:
maybe 🙃